crates/controller/src/hud/details.rs

```diff
@@ -1,37 +1,130 @@
 use bevy::prelude::*;
+use de_core::baseset::GameSet;
 use de_core::{cleanup::DespawnOnGameExit, gamestate::GameState};
+use de_energy::Battery;
+use de_gui::{BodyTextCommands, BodyTextOps, GuiCommands, OuterStyle};
 
 use super::{interaction::InteractionBlocker, HUD_COLOR};
+use crate::selection::Selected;
+
+const PREFIXES: [&str; 5] = ["T", "G", "M", "k", ""];
 
 pub(crate) struct DetailsPlugin;
 
 impl Plugin for DetailsPlugin {
     fn build(&self, app: &mut App) {
-        app.add_system(spawn_details.in_schedule(OnEnter(GameState::Playing)));
+        app.add_system(setup.in_schedule(OnEnter(GameState::Playing)))
+            .add_system(
+                update
+                    .in_base_set(GameSet::PostUpdate)
+                    .run_if(in_state(GameState::Playing)),
+            );
     }
 }
 
-fn spawn_details(mut commands: Commands) {
-    commands.spawn((
-        NodeBundle {
-            style: Style {
+#[derive(Resource)]
+struct DetailsText(Entity);
+
+fn setup(mut commands: GuiCommands) {
+    let node = commands
+        .spawn((
+            NodeBundle {
+                style: Style {
+                    size: Size {
+                        width: Val::Percent(20.),
+                        height: Val::Percent(30.),
+                    },
+                    position_type: PositionType::Absolute,
+                    position: UiRect::new(
+                        Val::Percent(0.),
+                        Val::Percent(20.),
+                        Val::Percent(70.),
+                        Val::Percent(100.),
+                    ),
+                    ..default()
+                },
+                background_color: HUD_COLOR.into(),
+                ..default()
+            },
+            DespawnOnGameExit,
+            InteractionBlocker,
+        ))
+        .id();
+    let details_text = commands
+        .spawn_body_text(
+            OuterStyle {
                 size: Size {
                     width: Val::Percent(20.),
                     height: Val::Percent(30.),
                 },
-                position_type: PositionType::Absolute,
-                position: UiRect::new(
-                    Val::Percent(0.),
-                    Val::Percent(20.),
-                    Val::Percent(70.),
-                    Val::Percent(100.),
-                ),
-                ..default()
+                margin: UiRect::all(Val::Percent(5.)),
             },
-            background_color: HUD_COLOR.into(),
-            ..default()
-        },
-        DespawnOnGameExit,
-        InteractionBlocker,
-    ));
+            "",
+        )
+        .id();
+    commands.entity(node).add_child(details_text);
+
+    commands.insert_resource(DetailsText(details_text));
+}
+
+fn format_units(value: f64, units: &str) -> String {
```

Comment:
I would write unnittests for this one: with at least 0, smaller than 0 and a few larger numbers.